### PR TITLE
Basic greedy assignment combination

### PR DESCRIPTION
**Pull request by Schaeff**
Instead of going through asm statements one by one, go through them as batches of compatible statements, and yield at most one code line per batch.

Currently only implemented disjoint assignments (different assignment registers and different assignment target)

### CODE REVIEW

**Comment by Schaeff** [comment]
@chriseth I started implementing this in `pilgen` based on the asm statements. However to decide which statements to combine, I need to do some analysis on the expression to be assigned, but pilgen uses the basic PIL AST not the analyzed one which has the relevant utils. Am I missing something? Should this be implemented somewhere else? Or should the analyzed AST be used in pilgen?

File: pilgen/src/lib.rs [hunk 1]
```diff
@@ -49,6 +51,109 @@ impl<T: FieldElement> ASMPILConverter<T> {
         Default::default()
     }
 
+    fn compatible(&self, left: &ASMStatement<T>, right: &ASMStatement<T>) -> bool {
```
Comment (chriseth) [inline pilgen/src/lib.rs:54]: Maybe split some stuff into a new file?

File: test_data/asm/disjoint_assignment_reg.asm [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+reg pc[@pc];
+reg X[<=];
+reg Y[<=];
+reg A;
+reg B;
+
+instr loop { pc' = pc }
+
+// the two following lines should be combined into a single one as they use disjoint registers
```
Comment (chriseth) [inline test_data/asm/disjoint_assignment_reg.asm:11]: Wait but how does it combine the free inputs?

Comment (chriseth) [reply test_data/asm/disjoint_assignment_reg.asm:11]: Oh it's different assignment registers!

File: pilgen/src/lib.rs (hunk 1 above)
Comment (Schaeff) [reply pilgen/src/lib.rs:54]: Yes I suspect this will not require `self` and can be on an Expression impl

File: parser/src/asm_ast.rs [hunk 3]
```diff
@@ -2,6 +2,52 @@ use number::AbstractNumberType;
 
 use super::ast::{Expression, SelectedExpressions, Statement};
 
+pub mod batched {
```
Comment (chriseth) [inline parser/src/asm_ast.rs:5]: Maybe the parser is not the right place, since this analyzes semantics

File: pilgen/src/batcher.rs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                    Expression::PolynomialReference(pol) => write_regs_1.contains(&pol.name),
+                    _ => false,
+                }) && !expr_any(value_1, |e: &Expression<T>| match e {
+                    Expression::PolynomialReference(pol) => write_regs_1.contains(&pol.name),
+                    _ => false,
+                }) {
+                } else {
+                    return Err(Incompatible::WriteAfterRead);
+                }
+                // if we are making function calls, we cannot use the pc // TODO allow a single static jump per batch
```
Comment (chriseth) [inline pilgen/src/batcher.rs:67]: I don't see the connection between function calls and the pc

File: parser/src/asm_ast.rs (hunk 3 above)
Comment (Schaeff) [reply parser/src/asm_ast.rs:5]: Right so by an analysis structure do you mean something like

```
struct ASMAnalysis {
   statements: Vec<ASMStatement>,
   batches: Option<Batches>
}
```

where the batcher sets the `batches` member and the pil converter optionally uses the batches?

File: pilgen/src/batcher.rs (hunk 4 above)
Comment (Schaeff) [reply pilgen/src/batcher.rs:67]: Changed to checking the whole expression

**Pull request closed without merging**